### PR TITLE
Gyro model

### PR DESCRIPTION
**Pull request by cameron-goddard**
### Summary

This PR addresses Jira ticket [CISLUNAR-165](https://jira.cornell.edu/browse/CISLUNAR-165).

`parameters.py` was updated with the gyro bias and gyro noise values. A new file `gyro_model.py` was created to implement the model as a subclass of `SensorModel`.

### Testing

Testing is not possible yet, as it is dependent on `state.py` being functional

### Notes

[Gyro Sensor Model](https://cornell.app.box.com/folder/158750635578)

### Comments
- [x] Add an x between the brackets if you commented your code well!

### CODE REVIEW

File: src/core/parameters.py [hunk 1]
```diff
@@ -8,8 +8,8 @@
 
 class Parameters:
     def __init__(self, param_dict: Dict = {}):
-        self.gyro_bias = 0
-        self.gyro_noise = 0
+        self.gyro_bias = [0.497625, -0.10821875, 0.77490625] #truncate to same # of decimal places?
```
Comment (tmf97) [inline src/core/parameters.py:11]: Don't worry, no need to truncate decimal places

File: src/core/models/gyro_model.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    def evaluate(self, state: State) -> Dict[str, Any]:
+        return super().evaluate(state)
+    
+    def d_state(self, state: State) -> Dict[str, Any]:
+        ang_vel_i = np.array([state.ang_vel_x, state.ang_vel_y, state.ang_vel_z])
+        gyro_bias = np.array(self._parameters.gyro_bias)
+
+        ang_vel_d = ang_vel_i + gyro_bias
+        ang_vel_d = np.random.normal(loc=ang_vel_d, scale=self._parameters.gyro_noise, size=3)
+        return {"ang_vel.x": ang_vel_d[0], "ang_vel.y": ang_vel_d[1], "ang_vel.z": ang_vel_d[2]} #return as quaternion rate?
```
Comment (tmf97) [inline src/core/models/gyro_model.py:19]: No need to convert to quaternion rates here. The gyro sensor spits out only angular velocity rates in x/y/z, so this is good.

File: src/core/models/gyro_model.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+class GyroModel(SensorModel):
+    """The gyro sensor model. This model applies the gyro bias and noise as specified in parameters.py"""
+
+    def __init__(self, parameters: Parameters) -> None:
+        super().__init__(parameters)
+    
+    def evaluate(self, state: State) -> Dict[str, Any]:
+        return super().evaluate(state)
+    
+    def d_state(self, state: State) -> Dict[str, Any]:
```
Comment (tmf97) [inline src/core/models/gyro_model.py:13]: We'll probably want to change the `SensorModel` base class a little bit, but that's for a future PR

File: src/core/state.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    "time" : 0,
+    "ang_vel_x" : 0.0,
+    "ang_vel_y" : 0.0,
+    "ang_vel_z" : 0.0,
+    "gnc_pos_x" : 0.0,
+    "gnc_pos_y" : 0.0,
+    "gnc_pos_z" : 0.0,
+    "device_gom_on" : True,
+    "device_gom_funct" : True
+}
```
Comment (tmf97) [inline src/core/state.py:14]: Solid start, but let's get on state.py in a new PR

File: src/core/models/gyro_model.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+from core.models.model import SensorModel
+import numpy as np
+
+class GyroModel(SensorModel):
+    """The gyro sensor model. This model applies the gyro bias and noise as specified in parameters.py"""
```
Comment (tmf97) [inline src/core/models/gyro_model.py:5]: Can we add the box link to the doc you made for this?

File: src/core/models/gyro_model.py (hunk 2 above)
Comment (tmf97) [inline src/core/models/gyro_model.py:19]: The _one_ small thing that's missing here is the quantization of the gyro rates. A perfect `float` is much more precise than values the gyro sensor is able to return. So after adding the noise, snap the value to the closest value allowed by the precision of the gyroscope

**Review by tmf97 — CHANGES_REQUESTED** [review]
Solid stuff, only small changes

File: src/core/models/gyro_model.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        ang_vel_i = np.array([state.ang_vel_x, state.ang_vel_y, state.ang_vel_z])
+        gyro_bias = np.array(self._parameters.gyro_bias)
+
+        ang_vel_d = ang_vel_i + gyro_bias
+        ang_vel_d = np.random.normal(loc=ang_vel_d, scale=self._parameters.gyro_noise, size=3)
+
+        for i in range(3):
+            ang_vel_d[i] = self._parameters.gyro_sensitivity*int(self._parameters.gyro_sensitivity/2 + vel/self._parameters.gyro_sensitivity)
+        
+        return {"ang_vel.x": ang_vel_d[0], "ang_vel.y": ang_vel_d[1], "ang_vel.z": ang_vel_d[2]}
```
Comment (tmf97) [inline src/core/models/gyro_model.py:27]: We'll probably want to change the type that this function returns, but OK for now

File: src/core/parameters.py [hunk 7]
```diff
@@ -8,8 +8,9 @@
 
 class Parameters:
     def __init__(self, param_dict: Dict = {}):
-        self.gyro_bias = 0
-        self.gyro_noise = 0
+        self.gyro_bias = [0.497625, -0.10821875, 0.77490625]
+        self.gyro_noise = [0.1824535, 0.11738579, 0.19192256]
+        self.gyro_sensitivity = 0.15625
```
Comment (tmf97) [inline src/core/parameters.py:13]: please make sure your units are consistent

**Review by tmf97 — CHANGES_REQUESTED** [review]

**Review by tmf97 — APPROVED** [review]

File: src/core/models/gyro_model.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+class GyroModel(SensorModel):
+    """This model applies the gyro bias and noise as specified in parameters.py"""
+
+    def __init__(self, parameters: Parameters) -> None:
+        super().__init__(parameters)
+    
+    def evaluate(self, state: State) -> Dict[str, Any]:
+        return super().evaluate(state)
+    
+    def d_state(self, state: State) -> Dict[str, Any]:
```
Comment (sjz38) [inline src/core/models/gyro_model.py:17]: Add docstring for return values

**Review by tmf97 — APPROVED** [review]

**Pull request merged**